src/main/java/com/ea/restaurant/service/impl/Oauth2ServiceImpl.java

```diff
@@ -0,0 +1,98 @@
+package com.ea.restaurant.service.impl;
+
+import com.ea.restaurant.constants.Oauth2.GranType;
+import com.ea.restaurant.constants.Status;
+import com.ea.restaurant.entities.AppAccessToken;
+import com.ea.restaurant.entities.AppClient;
+import com.ea.restaurant.entities.AppRefreshToken;
+import com.ea.restaurant.exceptions.BcryptException;
+import com.ea.restaurant.exceptions.EntityNotFoundException;
+import com.ea.restaurant.grpc.LoginClientResponse;
+import com.ea.restaurant.repository.AppAccessTokenRepository;
+import com.ea.restaurant.repository.AppClientRepository;
+import com.ea.restaurant.repository.AppClientScopeRepository;
+import com.ea.restaurant.repository.AppRefreshTokenRepository;
+import com.ea.restaurant.service.Oauth2Service;
+import com.ea.restaurant.utils.Oauth2Util;
+import com.nimbusds.jose.JOSEException;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.security.crypto.bcrypt.BCrypt;
+import org.springframework.stereotype.Service;
+
+@Service
+public class Oauth2ServiceImpl implements Oauth2Service {
+
+  private final String oauth2SecretKey;
+  private final AppClientRepository appClientRepository;
+  private final AppClientScopeRepository appClientScopeRepository;
+  private final AppRefreshTokenRepository appRefreshTokenRepository;
+  private final AppAccessTokenRepository appAccessTokenRepository;
+
+  public Oauth2ServiceImpl(
+      @Value("${oauth2.secret.key}") String oauth2SecretKey,
+      AppClientRepository appClientRepository,
+      AppClientScopeRepository appClientScopeRepository,
+      AppAccessTokenRepository appAccessTokenRepository,
+      AppRefreshTokenRepository appRefreshTokenRepository) {
+    this.oauth2SecretKey = oauth2SecretKey;
+    this.appClientRepository = appClientRepository;
+    this.appClientScopeRepository = appClientScopeRepository;
+    this.appAccessTokenRepository = appAccessTokenRepository;
+    this.appRefreshTokenRepository = appRefreshTokenRepository;
+  }
+
+  @Override
+  public LoginClientResponse loginClient(String clientId, String clientSecret)
+      throws JOSEException {
+    validateClientCredentials(clientId, clientSecret);
+    var client =
+        this.appClientRepository
+            .findByClientIdAndEntityStatus(clientId, Status.ACTIVE)
+            .orElseThrow(EntityNotFoundException::new);
+    var scopes =
+        this.appClientScopeRepository
+            .findByAppClientIdAndEntityStatus(client.getId(), Status.ACTIVE)
+            .orElseThrow(EntityNotFoundException::new);
+    var accessToken = Oauth2Util.buildClientCredentialsAccessToken(client, scopes, oauth2SecretKey);
+    var refreshToken =
+        Oauth2Util.buildClientCredentialsRefreshToken(client, scopes, oauth2SecretKey);
+
+    var persistedRefreshToken = createRefreshToken(refreshToken, client);
+    createAccessToken(accessToken, persistedRefreshToken);
+
+    return LoginClientResponse.newBuilder()
```

Comment:
Create a response class not related to rpc to return the data from this method

Reply:
Do this in the RPC layer instead, let's keep gRPC away from the data services

Reply:
Done.